xdsl/irdl/declarative_assembly_format_parser.py

```diff
@@ -291,11 +291,13 @@ def verify_results(self, seen_variables: set[str]):
             self.seen_result_types, self.op_def.results, strict=True
         ):
             if not result_type:
-                if not result_def.constr.can_infer(seen_variables):
+                can_infer_type = result_def.constr.can_infer(seen_variables)
+                can_infer_length = not isinstance(result_def, VariadicDef | OptionalDef)
```

Comment:
This doesn't look correct to me. The length could be inferred by a range variable constraint.

Reply:
Excellent, how would one go about doing that? apparently we have no examples in the code.

Reply:
Didn't test this so might have syntax errors but I believe the following should work
```
@irdl_op_definition
class RangeVarOp(IRDLOperation):
    name = "test.range_var"
    T: ClassVar = RangeVarConstraint("T", EqAttrConstraint(IntegerType(1)))
    ins = var_operand_def(T)
    outs = var_result_def(T)

    assembly_format = "$ins attr-dict"
```

Reply:
I added a test case for it, with a not implemented error. Would you like me to file an issue to track the implementation of length inference?

Reply:
I believe the length inference should already work in this case, am I incorrect?

Reply:
Exactly, it should be done on the lines you commented on, but the current code "implements" this by taking the length of `None` if `None` is a `list`, which is never the case. It should be possible to infer the information for the case you've provided, but I'm not aware of an implementation of a working mechanism in the code we have.

Reply:
Ok, I've tried to implement this and it seems you are right in that we don't currently have the means to make a working test.